### ISSUE TITLE
Non-Browser POST requests fail when CSRF protection enabled

### ISSUE THREAD

**Issue by adam-tylr** [openshift/jenkins-openshift-login-plugin#47]
**Updated Summary of my posts:**  CSRF filters are executed before `OpenShiftPermissionFilter` so there is no `Authentication` set, which the CSRF filter depends on.

I'm trying to use an OpenShift service account token to kick off a job with the Jenkins API. I have the token in an Authorization header and I am able to make GET requests successfully. When I try to POST  to the API, I get "403 No Valid crumb was included in the request". 

I have tried getting a crumb from /crumbIssuer and attaching it in a request header, but this gives the same error unless I also send a JSESSIONID cookie. Ideally, I shouldn't need a session to make this request. Based on [JENKINS-22474](https://issues.jenkins-ci.org/browse/JENKINS-22474), the convention seems to be to not require the crumb at all when using a token to make the request. 

After some research I found the hudson.security.csrf.CrumbExclusion class which says it allows plugins to define exceptions to the CSRF protection filter. I've been messing around with it but have not gotten anywhere.

**Comment by adam-tylr** [comment]
I've been approaching this from two different directions:

- **Why isn't the crumb from /crumbIssuer being accepted?**
Looking at `DefaultCrumbIssuer`, the crumb is generated using the `Authentication` name. When the crumb is generated, `authentication.getName()` returns my service account user ID. When I try to make the POST request to start the job however, it returns anonymous. The names don't match so it fails. It seems that the bearer token is ignored. In fact, I can send a blank token and I still get an invalid crumb response, not unauthorized. I added some logging to `OpenShiftPermissionFilter.java` and it seems like the authorization header is null when the filter is run. This is only for the POST request, not GET. I am able to start the job using a crumb from an anonymous request to /crumbIssuer but to do this I had to edit our security settings to allow anonymous access which I can't do for our production Jenkins. 
- **Can I disable the crumb check for token requests**
After looking at some other projects as examples I forked this repo and tried to implement a `CrumbExclusion` extension. For whatever reason, the extension is not being picked up by Jenkins. I'm not familiar with developing plugins so I'm not sure what's going wrong. I then tried to create a new plugin which just implements the exclusion extension and after installing I was able to make the POST. Developing our own plugin is probably not going to be a viable solution though.

**Comment by adam-tylr** [comment]
I figured out my issue implementing a `CrumbExclusion`. I was using Maven incorrectly. I did a blanket exclusion and it worked. Then I went back and tried to add logic to only exclude requests that have been authenticated with a Bearer token following the implementation for  `jenkins.security.ApiCrumbExclusion`. I ran into the same issue as with using the crumb though. It seems all the CSRF filters are executed before the `OpenShiftPermissionFilter` so there's no way for my crumb filter to make sure the user is passing a valid bearer token. I think I finally figured that out though. In the Jenkins web.xml, the filters are defined in the following order:

1.  `hudson.security.HudsonFilter`
2. `hudson.security.csrf.CrumbFilter`
3. `hudson.util.PluginServletFilter`

This plugin uses `hudson.util.PluginServletFilter` rather than `hudson.security.HudsonFilter`. Assuming the filters are executed in the order they're declared, there is no Authentication when the crumb filter runs. I see there's a comment in `OpenShiftOAuth2SecurityRealm` explaining this was done because of some other coupling issues.

At this point I'm not sure what the optimal solution would be. Is it sufficient to exclude CSRF protection based on the existence of the Bearer token in the request without actually validating it or does this require some other change?

**Comment by gabemontero** [comment]
Hey @adam-tylr 

The work I did to add token based authentication definitely predates https://issues.jenkins-ci.org/browse/JENKINS-22474 by over a year I believe.

I have not looked into the precise details of what that work changed, but it might have made using `hudson.security.HudsonFilter` or `hudson.security.SecurityRealm.createFilter` viable.

Did you in fact try changing the plugin to use that extension point and give it a try?

At this time, given our priorities/needs, I don't see my team investing time sort this out, but if you are interested in continuing to iterate on this, I can review results, provide feedback, etc.  If you get a solution that addresses both the CSRF and non-CSRF scenarios, we can merge it and cut a new version of the plugin.

thanks

**Comment by rajatsingh25aug** [comment]
I would love to figure out the issue but  @gabemontero  @adam-tylr   can u please provide steps to reproduce the code?.I'm not able to do that.
Thanks

**Comment by gabemontero** [comment]
Don't have the bandwidth to help you too much @rajatsingh25aug , but I can say from @adam-tylr 's comment on Nov 1:

- he must be switching https://github.com/openshift/jenkins-openshift-login-plugin/blob/master/src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftPermissionFilter.java#L66 from simply implementing Filter, to extending https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/security/HudsonFilter.java
- he must be changing https://github.com/openshift/jenkins-openshift-login-plugin/blob/master/src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java#L306-L322 to use https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/security/SecurityRealm.java#L468

Otherwise I'm assuming you know how to 
- clone this repo and build via mvn with a jdk installed
- go into the jenkins console and manually install hpi files for new versions of plugin
- run jenkins in an openshift cluster

If you don't know those items, then apologies, you'll have to research that on your own

thanks for the interest / willingness to help

**Comment by rajatsingh25aug** [comment]
Hey, @gabemontero thanks for the comment!
I don't know what hpi files are and how to install it but will try to learn about them and start on the issue very soon.
Is there's any slack channel where I can connect with you and you can help me out if I'll be needing help?.
You can mentor me in fixing this issue if you have time for it?.
Thanks

**Comment by gabemontero** [comment]
On Tue, Feb 19, 2019 at 1:26 PM Rajat Singh <notifications@github.com>
wrote:

> Hey, @gabemontero <https://github.com/gabemontero> thanks for the comment!
> I don't know what hpi files are and how to install it but will try to
> learn about them and start on the issue very soon.
> Is there's any slack channel where I can connect with you and you can help
> me out if I'll be needing help?.
>
Sorry, no.  Any exchanges would have to be through this issue.

> You can mentor me in fixing this issue if you have time for it?.
>
I'll  try but my time will be very limited, since this change is not
prioritized for me.

**Comment by rajatsingh25aug** [comment]
Thanks :+1: 
I'll try to figure it out by myself and will post any doubts here!

**Comment by rajatsingh25aug** [comment]
> Don't have the bandwidth to help you too much @rajatsingh25aug , but I can say from @adam-tylr 's comment on Nov 1:
> 
> * he must be switching https://github.com/openshift/jenkins-openshift-login-plugin/blob/master/src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftPermissionFilter.java#L66 from simply implementing Filter, to extending https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/security/HudsonFilter.java
> * he must be changing https://github.com/openshift/jenkins-openshift-login-plugin/blob/master/src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java#L306-L322 to use https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/security/SecurityRealm.java#L468
> 
> Otherwise I'm assuming you know how to
> 
> * clone this repo and build via mvn with a jdk installed
> * go into the jenkins console and manually install hpi files for new versions of plugin
> * run jenkins in an openshift cluster
> 
> If you don't know those items, then apologies, you'll have to research that on your own
> 
> thanks for the interest / willingness to help

I followed your steps and was able to reproduce the error @gabemontero . :100: 
![screenshot from 2019-03-04 17-25-59](https://user-images.githubusercontent.com/17253992/53732285-7f6dbb80-3ea3-11e9-975d-6d870fd383ad.png)
But as @adam-tylr  mentioned that he attached the JSESSIONID in the header to execute the command successfully.
Can you please tell me in how can I get the JSESSIONID?.

**Comment by adam-tylr** [comment]
@rajatsingh25aug I believe the JSESSIONID defeats the purpose. The JSESSIONID is a cookie that's used to reference the user's session. It's created by Jenkins when you make the first request. If using the service account token, the authentication should come from the token being passed in every request. The client shouldn't need to also pass the session cookie.

**Comment by rajatsingh25aug** [comment]
@adam-tylr as you have mentioned in your issue that you also passed the JSESSIONID in order to send the request successfully. Right now I'm trying to investigate the issue and trying to find out what is causing this error.
Since you have mentioned that when you have passed the TOKEN, JSESSIONID as a header in that POST request, then you were able to make the curl request easily.
Can you attach some screenshots so that I can see what JSESSIONID cookie you are using?.
Because I'm not able to find it, neither am I able to find the location where Jenkins is creating the JSESSION.

**Comment by adam-tylr** [comment]
Sure, here are the steps:

1. Make a random request to Jenkins with the token in the authorization header. 
```
$ curl -X GET   https://jenkins.mycompany.com/job/DUMMY-JOB/build   
    -H 'Authorization: Bearer <token>'   
    -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b'
    -H 'cache-control: no-cache' -i
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
HTTP/1.1 405 Method Not Allowed
Date: Tue, 05 Mar 2019 02:34:35 GMT
X-Content-Type-Options: nosniff
Allow: POST
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Cache-Control: no-cache,no-store,must-revalidate
X-Hudson-Theme: default
Referrer-Policy: same-origin
Content-Type: text/html;charset=utf-8
Set-Cookie: JSESSIONID.242f0932=node0zfguod3k5dbo19sjnms115rdu696653.node0;Path=/;Secure;HttpOnly
Content-Length: 11449
Server: Jetty(9.4.z-SNAPSHOT)
Set-Cookie: b3c0f76d3ddd21235737fbe22847a1b6=d248ee83838bfe83d48ea3334b7a421b; path=/; HttpOnly; Secure

<resonse body>
```
Notice the Set-Cookie: JSESSIONID... header in the response. This is the cookie that you include with later calls.
2. Make a call to Jenkins to get the crumb
```
$ curl -X GET \
>   https://jenkins.mycompany.com/crumbIssuer/api/json \
>   -H 'Authorization: Bearer <token>' \
>   -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b' \
>   -H 'cache-control: no-cache'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   131  100   131    0     0    139      0 --:--:-- --:--:-- --:--:--   147
{"_class":"hudson.security.csrf.DefaultCrumbIssuer","crumb":"238e4690e619ee1260e49f416c0ec88b","crumbRequestField":"Jenkins-Crumb"}
```
3. Make a POST request with the crumb to start a job. Note that in this request we don't even have to send a bearer token since we have the session, hence it defeats the purpose.
```
$ curl -X POST --cookie "JSESSIONID.242f0932=node0zfguod3k5dbo19sjnms115rdu696653.node0" \
>   https://jenkins.mycompany.com/job/DUMMY-JOB/build \
>   -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b' \
>   -H 'cache-control: no-cache'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

```

**Comment by rajatsingh25aug** [comment]
> Sure, here are the steps:
> 
> 1. Make a random request to Jenkins with the token in the authorization header.
> 
> ```
> $ curl -X GET   https://jenkins.mycompany.com/job/DUMMY-JOB/build   
>     -H 'Authorization: Bearer <token>'   
>     -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b'
>     -H 'cache-control: no-cache' -i
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> HTTP/1.1 405 Method Not Allowed
> Date: Tue, 05 Mar 2019 02:34:35 GMT
> X-Content-Type-Options: nosniff
> Allow: POST
> Expires: Thu, 01 Jan 1970 00:00:00 GMT
> Cache-Control: no-cache,no-store,must-revalidate
> X-Hudson-Theme: default
> Referrer-Policy: same-origin
> Content-Type: text/html;charset=utf-8
> Set-Cookie: JSESSIONID.242f0932=node0zfguod3k5dbo19sjnms115rdu696653.node0;Path=/;Secure;HttpOnly
> Content-Length: 11449
> Server: Jetty(9.4.z-SNAPSHOT)
> Set-Cookie: b3c0f76d3ddd21235737fbe22847a1b6=d248ee83838bfe83d48ea3334b7a421b; path=/; HttpOnly; Secure
> 
> <resonse body>
> ```
> Notice the Set-Cookie: JSESSIONID... header in the response. This is the cookie that you include with later calls.
> 2. Make a call to Jenkins to get the crumb
> 
> ```
> $ curl -X GET \
> >   https://jenkins.mycompany.com/crumbIssuer/api/json \
> >   -H 'Authorization: Bearer <token>' \
> >   -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b' \
> >   -H 'cache-control: no-cache'
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100   131  100   131    0     0    139      0 --:--:-- --:--:-- --:--:--   147
> {"_class":"hudson.security.csrf.DefaultCrumbIssuer","crumb":"238e4690e619ee1260e49f416c0ec88b","crumbRequestField":"Jenkins-Crumb"}
> ```
When you have the crumb, why are you making a request to get the crumb again?
> 1. Make a POST request with the crumb to start a job. Note that in this request we don't even have to send a bearer token since we have the session, hence it defeats the purpose.
> 
> ```
> $ curl -X POST --cookie "JSESSIONID.242f0932=node0zfguod3k5dbo19sjnms115rdu696653.node0" \
> >   https://jenkins.mycompany.com/job/DUMMY-JOB/build \
> >   -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b' \
> >   -H 'cache-control: no-cache'
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> ```

**Comment by rajatsingh25aug** [comment]
Hey @adam-tylr @gabemontero .After investigating the issue from my knowledge.This is what I have to conclude.

- When CSRF protection is enabled and once we have the JSESSIONID , then the bearer token is not serving it's purpose.

- We have to find a way to check why the **Bearer Token** is not able to full fill its purpose.

So that means somewhere in the code, either of the two things is happening.

- Token value is getting generated but not getting consumed and that is why once we have the JSESSIONID, the token is not needed.

Since @gabemontero  is the maintainer of this plugin, I wanted to know if you can help me in from which file can I start making changes in the code. Since the codebase for this plugin is so huge for me as I'm new to the openSource.
Any help/leads will be appreciated.

**Comment by rajatsingh25aug** [comment]
> Sure, here are the steps:
> 
> 1. Make a random request to Jenkins with the token in the authorization header. 
> Are you making these requests with CSRF enabled or disabled.Please have a look at my comment that I posed earlier with the screenshot, I did that curl request with CSRF enabled.
> ```
> $ curl -X GET   https://jenkins.mycompany.com/job/DUMMY-JOB/build   
>     -H 'Authorization: Bearer <token>'   
>     -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b'
>     -H 'cache-control: no-cache' -i
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> HTTP/1.1 405 Method Not Allowed
> Date: Tue, 05 Mar 2019 02:34:35 GMT
> X-Content-Type-Options: nosniff
> Allow: POST
> Expires: Thu, 01 Jan 1970 00:00:00 GMT
> Cache-Control: no-cache,no-store,must-revalidate
> X-Hudson-Theme: default
> Referrer-Policy: same-origin
> Content-Type: text/html;charset=utf-8
> Set-Cookie: JSESSIONID.242f0932=node0zfguod3k5dbo19sjnms115rdu696653.node0;Path=/;Secure;HttpOnly
> Content-Length: 11449
> Server: Jetty(9.4.z-SNAPSHOT)
> Set-Cookie: b3c0f76d3ddd21235737fbe22847a1b6=d248ee83838bfe83d48ea3334b7a421b; path=/; HttpOnly; Secure
> 
> <resonse body>
> ```
> Notice the Set-Cookie: JSESSIONID... header in the response. This is the cookie that you include with later calls.
> 2. Make a call to Jenkins to get the crumb
> 
> ```
> $ curl -X GET \
> >   https://jenkins.mycompany.com/crumbIssuer/api/json \
> >   -H 'Authorization: Bearer <token>' \
> >   -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b' \
> >   -H 'cache-control: no-cache'
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100   131  100   131    0     0    139      0 --:--:-- --:--:-- --:--:--   147
> {"_class":"hudson.security.csrf.DefaultCrumbIssuer","crumb":"238e4690e619ee1260e49f416c0ec88b","crumbRequestField":"Jenkins-Crumb"}
> ```
> 1. Make a POST request with the crumb to start a job. Note that in this request we don't even have to send a bearer token since we have the session, hence it defeats the purpose.
> This command gives the me that same 403 error as my CSRF is enabled.I don't know if you have your CSRF enabled or disabled.
> ```
> $ curl -X POST --cookie "JSESSIONID.242f0932=node0zfguod3k5dbo19sjnms115rdu696653.node0" \
> >   https://jenkins.mycompany.com/job/DUMMY-JOB/build \
> >   -H 'Jenkins-Crumb: 238e4690e619ee1260e49f416c0ec88b' \
> >   -H 'cache-control: no-cache'
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> ```

**Comment by rajatsingh25aug** [comment]
Hey @adam-tylr @gabemontero .After investigating the issue from my knowledge.This is what I have to conclude.

- When CSRF protection is enabled and once we have the JSESSIONID , then the bearer token is not serving it's purpose.

- We have to find a way to check why the **Bearer Token** is not able to full fill its purpose.

So that means somewhere in the code, either of the two things is happening.

- Token value is getting generated but not getting consumed and that is why once we have the JSESSIONID, the token is not needed.

Any help/leads will be appreciated.

**Comment by adam-tylr** [comment]
@rajatsingh25aug The issue is the order of the servlet filters and the filter used by this plugin. You can see in the Jenkins web.xml that they will be executed in this order:
1. `hudson.security.HudsonFilter`
2. `hudson.security.csrf.CrumbFilter`
3. `hudson.util.PluginServletFilter` 

This plugin uses uses `hudson.util.PluginServletFilter`. So when the crumbFilter is run, the current user is anonymous. It's not until the next step that the token is read and authentication created. The JSESSIONID eliminates this issue because the authentication is created earlier in the filter chain based off that cookie. Plugins have the option to implement `hudson.util.PluginServletFilter` or  `hudson.security.HudsonFilter`. It seems like the solution would be to switch to `hudson.security.HudsonFilter` so that the authentication is done before the crumb verification. This is what I was trying to do back in November but simply switching them out created other issues that I didn't have the bandwidth to debug.

**Comment by gabemontero** [comment]
Yeah @rajatsingh25aug I agree with @adam-tylr synopsis.

Although it has been quite a while since I originally added the non-browser support, per what I put in the comment at https://github.com/openshift/jenkins-openshift-login-plugin/blob/master/src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java#L299-L304 there are I am assuming there is still a fair amount of the tight coupling and ordering of things that I alluded to there, and @adam-tylr notes on those specifics above were at least part of what I was referring to.

Also, historically speaking, we were leery of enabling CSRF out of the box because of the gotchas noted at https://wiki.jenkins.io/display/JENKINS/CSRF+Protection

The main motivation for the non-browser access was to facilitate such scripts. 

My two bottom line takeaways:
1) I would not be surprised if we would have to extend and replace the jenkins instance default crumb issuer and at its issueCrumb method at https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/security/csrf/DefaultCrumbIssuer.java#L73-L92 augment the call to `Jenkins.getAuthentication` to leverage our oauth flow based on the token provided ... the Jenkins instance does have a setCrumbIssuer method. ... a tricky endeavour given the jenkins "security" landscape's fluidity.

2) Aside from folks like @adam-tylr sorting it out for us, assuming @rajatsingh25aug you are on Pradeepto's team under the devtools org, aside from this being an educational exercise I at least would still consider this lower priority to all the other changes and things to learn that you'll soon have on your plate.

**Comment by rajatsingh25aug** [comment]
Hey @gabemontero When I logged `**token**` here https://github.com/openshift/jenkins-openshift-login-plugin/blob/master/src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftPermissionFilter.java#L187  and made a `**GET**` request o here , It showed me the token that I sent via header. But when I tried the same with the `**POST**` request, it logged nothing. I think the issue is not with the plugin.It's the jenkins that is not picking the header values when sent as a `**POST**` but working fine when making a **GET** request.I may be wrong, but I'm writing here what I tested.Can we do something about it?.I'm not famaliar with the Jenkins codebase and don't know where to look at to find this issue.

**Comment by rajatsingh25aug** [comment]
@adam-tylr  What version of Jenkins are you using? And have you tried other versions for Jenkins for the same issue?

**Comment by adam-tylr** [comment]
I'm on 2.150.1 and have not tried any other versions

**Comment by gabemontero** [comment]
@rajatsingh25aug - in our current e2e tests we are making POSTs to the jenkins server today with the token (when csrf is not enabled of course) and it works.

see https://github.com/openshift/origin/blob/master/test/extended/builds/pipeline_jenkins_e2e.go#L523

I would not expect that POST to work if we were not authenticated.

The logic behind that POST call is in https://github.com/openshift/origin/blob/master/test/extended/util/jenkins/ref.go

As an educational exercise, and probably of more immediate short term value for you from my perspective than implementing CSRF, would be for you to bring up the bluegreen sample from that e2e test case with your debug and proving what happens with the POST there.

**Comment by bsquizz** [comment]
Running into this as well. I was trying to use the Parametrized Remote Trigger Plugin to trigger/monitor builds on a remote Jenkins which has Open Shift auth enabled. Couldn't get the plugin to work properly until I disabled CSRF checks.

**Comment by foch** [comment]
We had the same issue. To have Jenkins crumb and Service Account tokens live together, we are using the JSESSIONID hack.
The trick is to create a JSESSIONID by curling Jenkins root URL, then create a crumb with this JSESSIONID, and subsequent calls to Jenkins API can be done by providing both the crumb and the Service Account token. 
Note that it fails if we keep providing the JSESSIONID cookie o__O

Working solution here: https://gist.github.com/foch/c4fe1141cf5baf73e42e014a8beb0d64

**Comment by adam-tylr** [comment]
@foch If you're interested in why the last step fails if you include the cookie:
The cookie corresponds to an anonymous authentication. When you make the POST request to start a build, the request hits the crumb verifier before it hits the openshift login plugin. So no crumb means it never even tries to read the token. Your solution works because the crumb you include is valid for an anonymous user so it passes the verifier step and makes it to the openshift login to create the authentication from the token. If you include the cookie in that step then you are already authenticated as anonymous and are not authorized to start the build. The token is ignored.

So your solution will work but only if you can control how cookies are used. In my case, I'm using a client library so I have no control over that.

**Comment by akram** [comment]
based on the proposed workaround here
Working solution here: https://gist.github.com/foch/c4fe1141cf5baf73e42e014a8beb0d64

and that a final fix would require to modify web.xml in jenkins war, we are closing this issue.

/close

**Comment by svenstaro** [comment]
Sadly, for me the workaround doesn't work. I get:

```
OK - Token for system:serviceaccount:redacted:jenkinstest retrieved
OK - Jenkins JSESSIONID generated (JSESSIONID.c111160=node01qkl8nrajsdade174.node0)
OK - Jenkins crumb successfully generated (Jenkins-Crumb:7ce3e1e2ed611dadasdsa41fdasf1232f94e789146909b00a53aae4dc)
<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
<title>Error 403 No valid crumb was included in the request</title>
</head>
<body><h2>HTTP ERROR 403</h2>
<p>Problem accessing //job/Administration/job/Utils/job/test/build. Reason:
<pre>    No valid crumb was included in the request</pre></p><hr><a href="http://eclipse.org/jetty">Powered by Jetty:// 9.4.z-SNAPSHOT</a><hr/>

</body>
</html>
```

Crumb security and proxy compatibility flags are enabled.

**Comment by foch** [comment]
@svenstaro When migrating to a more recent version of Jenkins I had to add the startup flag:

`-Dhudson.security.csrf.DefaultCrumbIssuer.EXCLUDE_SESSION_ID=true`

**Comment by a1dutch** [comment]
Hi,

I faced the same problem as above with Jenkins 2.235.1 after an update, it is no longer possible to disable csrf protection and the Jenkins-Crumb does not match which is causing it to fail.

I did some debugging and you can see the crumbs generated below with bearer auth provided, first generated by calling the `/crumbIssuer/api/json` endpoint and the second one of my jobs with the generated crumb.

`system:serviceaccount:test:jenkins-view-edit;192.168.0.2;nodeXXXXXXXXXXXX, salt: xxxxxxxxxxxxxxxx`

`anonymous;192.168.0.2;nodeXXXXXXXXXXXX, salt: xxxxxxxxxxxxxxxx`

I looked through the code and tried to setup an environment out side of my cluster but can not get this to work (cant set the namespace parameter to look for the config map).

It seems the `OpenShiftPermissionFilter` should be pulling the `sub` field out of the jwt and setting it on the `Authentication.getName()` which would then make the crumb valid.

Maybe someone with a better understanding of the code could point me in the right direction to fix this?

Cheers

@a1dutch

**Comment by adam-tylr** [comment]
@a1dutch The workaround stopped working for us when upgrading too and as @foch said you can revert to the old crumb issuer with `-Dhudson.security.csrf.DefaultCrumbIssuer.EXCLUDE_SESSION_ID=true` and it will still work. Although the option has been removed from the UI, you can also still disable CSRF protection by setting `-Dhudson.security.csrf.GlobalCrumbIssuerConfiguration.DISABLE_CSRF_PROTECTION=true` but in the release notes they warn this option may be removed in the future

**Comment by a1dutch** [comment]
@adam-tylr thanks I never had seen that option in the release notes to disable the crumb issuer.

Although it's a work around it doesn't fix the underlying issue with the crumb not matching due to the name being anonymous.
